File: DesignPatterns/src/Java/002-GOF-Patterns/002-StructuralPatterns/001-Composite/ShapeGroup/src/org/csystem/app/ShapeGroup.java
package org.csystem.app;

import java.util.ArrayList;
import java.util.List;

public class ShapeGroup implements IShape {
    private List<IShape> m_shapes = new ArrayList<>();

    public ShapeGroup addShape(IShape shape)
    {
        m_shapes.add(shape);

        return this;
    }

    public IShape getShape(int index)
    {
        return m_shapes.get(index);
    }

    public int size()
    {
        return m_shapes.size();
    }

    @Override
    public void draw()
    {
        System.out.println("Shapegroup.draw");
        for (var shape : m_shapes)
            shape.draw();
    }
}

File: DesignPatterns/src/C++/002-GOFPatterns/002-StructuralPatterns/001-Composite/ShapeGroup/ShapeGroup.cpp
#include "ShapeGroup.h"

#include <iostream>
#include <algorithm>


using namespace std;

void ShapeGroup::Draw()
{
	cout << "ShapeGroup::Draw()" << endl;

	std::for_each(m_shapes.begin(), m_shapes.end(), [](auto s) {s->Draw();});
}

void ShapeGroup::AddShape(Shape *shape)
{
	m_shapes.push_back(shape);
}